File: core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/Direction.kt
package fr.sncf.osrd.utils

import fr.sncf.osrd.railjson.schema.common.graph.EdgeDirection

/** A direction along an axis */
enum class Direction {
    /** The direction along which the position measure increases */
    INCREASING,
    /** The direction along which the position measure decreases */
    DECREASING;

    val opposite: Direction
        get() =
            when (this) {
                INCREASING -> DECREASING
                DECREASING -> INCREASING
            }

    val toEndpoint: Endpoint
        get() =
            when (this) {
                INCREASING -> Endpoint.END
                DECREASING -> Endpoint.START
            }

    /** Returns 1 for INCREASING, -1 for DECREASING */
    val sign
        get() =
            when (this) {
                INCREASING -> 1
                DECREASING -> -1
            }
}

fun EdgeDirection.toDirection(): Direction {
    return when (this) {
        EdgeDirection.START_TO_STOP -> Direction.INCREASING
        EdgeDirection.STOP_TO_START -> Direction.DECREASING
    }
}

File: python/railjson_generator/railjson_generator/schema/infra/direction.py
from enum import IntEnum


class Direction(IntEnum):
    START_TO_STOP = 0
    STOP_TO_START = 1

    def opposite(self):
        if self == Direction.START_TO_STOP:
            return Direction.STOP_TO_START
        return Direction.START_TO_STOP


class ApplicableDirection(IntEnum):
    START_TO_STOP = 0
    STOP_TO_START = 1
    BOTH = 2

    def directions(self):
        if self != ApplicableDirection.STOP_TO_START:
            yield Direction.START_TO_STOP
        if self != ApplicableDirection.START_TO_STOP:
            yield Direction.STOP_TO_START
